Add Limber MIX Mock Server and MIX Example

tarun018 opened the pull request:
Standard output shows:

Connecting...Connected
Request list of channels.
List of rooms at example.tg
	1. coven - coven@example.tg

Request supported nodes for MIX channel: coven@example.tg
Nodes supported by channel coven@example.tg
	1. urn:xmpp:mix:nodes:participants
	2. urn:xmpp:mix:nodes:messages
	3. urn:xmpp:mix:nodes:presence
	4. urn:xmpp:mix:nodes:jidmap

Successfully joined channel coven@example.tg
[warning] Swiften/MIX/MIXRegistry.cpp:24 joinChannel: Channel already joined: coven@example.tg
Successfully left channel coven@example.tg


Warning is intentional to notify about asking to join an already joined channel.

intosi commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,219 @@
+            channels_.insert(channelJID);
+        }
+
+        std::unordered_set<std::string>& getChannels() {
```
Shouldn't this be a const reference?

intosi commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,219 @@
+
+            //If request comes to domain service
+            if (stanza->getTo().isValid() && stanza->getTo() == session->getLocalJID()) {
+                if (std::shared_ptr<IQ> iq = std::dynamic_pointer_cast<IQ>(stanza)) {
```
Are you masking the existing iq variable with the same contents deliberately here?

intosi commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,219 @@
+                if(iq->getPayload<RosterPayload>()) {
+                    SWIFT_LOG(debug) << "Query: Roster" << std::endl;
+                    session->sendElement(IQ::createResult(iq->getFrom(), iq->getID(), std::make_shared<RosterPayload>()));
+                }
```
Continuing processing the rest of handleElementReceived after this sendElement (and the ones below) isn't really useful, is it?

intosi commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,219 @@
+    SimpleEventLoop eventLoop;
+    SimpleUserRegistry userRegistry;
+
+    userRegistry.addUser(JID("some@example.tg"), "mixtest");
```
I think I would prefer it if you used example.com, which is guaranteed to never mask existing domains, instead of assuming there will never be a .tg TLD.

intosi commented on Swiften/Examples/MIXListAndJoin/MIXListAndJoin.cpp:
```diff
@@ -0,0 +1,183 @@
+    assert(mixRegistry_);
+    assert(mixRegistry_->getMIXInstance(jid));
+
+    cout << "Successfully joined channel " << jid << std::endl;
```
Mixing prefixed and non-prefixed iostream bits looks odd.  I think I'd prefer the prefixed ones here.

intosi commented on Swiften/Roster/XMPPRosterImpl.cpp:
```diff
@@ -40,6 +40,15 @@ void XMPPRosterImpl::clear() {
 }
 
+bool XMPPRosterImpl::isMIXChannel(const JID& jid) {
+    std::map<JID, XMPPRosterItem>::const_iterator i = entries_.find(jid.toBare());
```
Assuming you really need the explicit const iterator here, I would prefer RosterMap::const_iterator instead.

tfar commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,219 @@
+            if (iq) {
+                if(iq->getPayload<RosterPayload>()) {
+                    SWIFT_LOG(debug) << "Query: Roster" << std::endl;
+                    session->sendElement(IQ::createResult(iq->getFrom(), iq->getID(), std::make_shared<RosterPayload>()));
```
This isn't the final version right? I would expect it to return the MIX channels which the user has joined.

tfar commented on Swiften/Client/Client.h:
```diff
@@ -191,6 +196,7 @@ namespace Swift {
             PresenceOracle* presenceOracle;
             DirectedPresenceSender* directedPresenceSender;
             StanzaChannelPresenceSender* stanzaChannelPresenceSender;
+            MIXRegistry* mixRegistry;
```
This could be a `unique_ptr`, right?

tarun018 commented:
Updated based on feedback and added roster sync between clients:

Example (Client 1) :
./Swiften/Examples/MIXListAndJoin/MIXListAndJoin another@example.com/some mixtest2 example.com
Connecting...Connected
Request list of channels.
List of rooms at example.com
	1. coven - coven@example.com

Request supported nodes for MIX channel: coven@example.com
Nodes supported by channel coven@example.com
	1. urn:xmpp:mix:nodes:participants
	2. urn:xmpp:mix:nodes:messages
	3. urn:xmpp:mix:nodes:presence
	4. urn:xmpp:mix:nodes:jidmap

Already Joined channels: 
No channels already joined.

Successfully joined channel coven@example.com
[warning] Swiften/MIX/MIXRegistry.cpp:40 joinChannel: Channel already joined: coven@example.com


However (client 2 of the same user):
./Swiften/Examples/MIXListAndJoin/MIXListAndJoin another@example.com/someelse mixtest2 example.com
Connecting...Connected
Request list of channels.
List of rooms at example.com
	1. coven - coven@example.com

Request supported nodes for MIX channel: coven@example.com
Nodes supported by channel coven@example.com
	1. urn:xmpp:mix:nodes:participants
	2. urn:xmpp:mix:nodes:messages
	3. urn:xmpp:mix:nodes:presence
	4. urn:xmpp:mix:nodes:jidmap

Already Joined channels: 
	1. coven - coven@example.com

[warning] Swiften/MIX/MIXRegistry.cpp:40 joinChannel: Channel already joined: coven@example.com

tfar commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,263 @@
+            stanza->setFrom(session->getRemoteJID());
+
+            auto iq = std::dynamic_pointer_cast<IQ>(stanza);
+            if (iq) {
```
It probably makes more sense to test for `!iq` and handle the error case there and afterwards check if it's a roster payload.

tarun018 replied:
Yeah, makes more sense.

tfar commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,263 @@
+            }
+
+            //If request comes to domain service
+            if (stanza->getTo().isValid() && stanza->getTo() == session->getLocalJID()) {
```
Is the `isValid()` test required here?

tfar commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,263 @@
+                }
+                return;
+            }
+        }
```
This method is quite long. It probably makes sense to split it up in parts. Like `handleElementReceivedForAccount`, `handleElementReceivedForMIXService`, `handleElementReceivedForMIXChannel`, etc.

tarun018 replied:
I've split the function into the three cases now.

tfar commented on Swiften/Elements/RosterPayload.h:
```diff
@@ -42,7 +42,16 @@ namespace Swift {
                 version_ = version;
             }
 
+            bool hasMIXAnnotationSupport() const {
```
This method name doesn't make it clear what it is for. You probably want to enable requesting MIX annotations for roster requests. How about `hasRequestMIXAnnotations()`? Similar for the following method.

tarun018 replied:
Updated.

tarun018 commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,263 @@
+ * All rights reserved.
+ * See the COPYING file for more information.
+ */
+/*
```
Will udpate.

tarun018 commented:
Updated based on feedback.

tarun018 commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,275 @@
+
+            auto iq = std::dynamic_pointer_cast<IQ>(stanza);
+            if (!iq) {
+                session->sendElement(IQ::createError(iq->getFrom(), iq->getID(), ErrorPayload::BadRequest, ErrorPayload::Cancel));
```
I am now returning if element is not an iq.

tarun018 commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,275 @@
+            }
+
+            //If request comes to domain service
+            if (stanza->getTo() == session->getLocalJID()) {
```
Updated.

tarun018 commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,275 @@
+
+        void handleElementReceivedForAccount(IQ::ref iq, std::shared_ptr<ServerFromClientSession> session) {
+            if (auto incomingJoinPayload = iq->getPayload<MIXJoin>()) {
+                if (auto channelJID = incomingJoinPayload->getChannel()) {
```
Added handleJoin and handleLeave methods.

tfar commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,277 @@
+                return;
+            }
+
+            if(iq->getPayload<RosterPayload>()) {
```
Shouldn't this if be in the `handleElementReceivedForAccount` method? After all roster queries go to the account, not?

tarun018 replied:
Roster queries does not have a "to" field, therefore if conditions down there won't pass.

tarun018 replied:
handleElementReceivedForAccount is handling all queries for user's own local server.

tfar commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,277 @@
+                handleElementReceivedForMIXChannel(iq, session);
+            }
+
+            //If request comes to domain service
```
Could you move the comments for these if inside the if block so that the `else if (…` directly follows the line with the `}`, thanks. The current style makes it appear an if statement has ended while the corresponding else is following up a couple lines below.

tfar commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,277 @@
+
+    private:
+        IDGenerator idGenerator_;
+        PlatformXMLParserFactory xmlParserFactory;
```
Inconsistent, should end with `_` as well.

tfar commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,274 @@
+                //If request comes to domain service
+                handleElementReceivedForMIXService(iq, session);
+            } else if (stanza->getTo() == session->getRemoteJID().toBare() || stanza->getTo() == JID()) {
+                //If request comes to own local server
```
This is to the account and not to the local server, the local server would be `to='server domain'`.

tarun018 replied:
Updated the comment.

tfar commented on Limber/MIXMockServer.cpp:
```diff
@@ -0,0 +1,274 @@
+        }
+
+        std::shared_ptr<RosterPayload> getRosterPayloadFromXMPPRoster(std::shared_ptr<XMPPRosterImpl> roster) {
+            auto rosterPayload = std::make_shared<RosterPayload>();
```
`getFoo()` usually describes a getter method. This seems to be a) not a method to retrieve a member and b) a method that does not otherwise access members, so can be static, right?

tarun018 replied:
Yes this can be a static method.

tfar commented on Swiften/Examples/MIXListAndJoin/MIXListAndJoin.cpp:
```diff
@@ -0,0 +1,201 @@
+
+#include <iostream>
+#include <memory>
+#include <unordered_set>
```
You are using `assert()` but do not `#include` cassert header.

tarun018 replied:
Included.

tfar commented on Swiften/Examples/MIXListAndJoin/MIXListAndJoin.cpp:
```diff
@@ -0,0 +1,201 @@
+    int channelCount = 0;
+    std::cout << "Already Joined channels: " << std::endl;
+    for (auto item : xmppRoster->getItems()) {
+        channelCount++;
```
Channels are only those in the roster that have the *mix* annotation. However, the MIX registry should probably be used to fetch the already joined channels, which has access to the roster. This way if the user wants to do something MIX related, they can always go tho the `MIXRegistry`, right? Opinions, @tarun018  @Kev ?

tarun018 replied:
Seems like a sensible idea. I wanted to hold my client till onInitialRosterPopulated is called and syncRegistryWithRoster is successfully executed.

tfar commented on Swiften/Examples/MIXListAndJoin/MIXListAndJoin.cpp:
```diff
@@ -0,0 +1,201 @@
+    mixRegistry_->onChannelJoinFailed.connect(&handleChannelJoinFailed);
+    mixRegistry_->onChannelLeft.connect(&handleChannelLeft);
+
+    XMPPRoster* xmppRoster = client->getRoster();
```
Please use `auto` here.

tfar commented on Swiften/Examples/MIXListAndJoin/MIXListAndJoin.cpp:
```diff
@@ -0,0 +1,201 @@
+}
+
+static void handleRosterPopulated() {
+    XMPPRoster* xmppRoster = client->getRoster();
```
Please use `auto` here.

tfar commented on Swiften/Examples/MIXListAndJoin/MIXListAndJoin.cpp:
```diff
@@ -0,0 +1,201 @@
+        std::cout << "Connecting..." << std::flush;
+        client->connect(options);
+        {
+            Timer::ref timer = networkFactories.getTimerFactory()->createTimer(30000);
```
Please use `auto´ here.

tfar commented on Swiften/Examples/MIXListAndJoin/MIXListAndJoin.cpp:
```diff
@@ -0,0 +1,201 @@
+            Timer::ref timer = networkFactories.getTimerFactory()->createTimer(30000);
+            timer->onTick.connect(boost::bind(&SimpleEventLoop::stop, &eventLoop));
+
+            Timer::ref disconnectTimer = networkFactories.getTimerFactory()->createTimer(25000);
```
Please use `auto` here.

tfar commented on Swiften/MIX/MIXRegistry.h:
```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include <memory>
+#include <unordered_set>
```
Missing a `#include <string>`.

tfar commented on Swiften/Examples/MIXListAndJoin/MIXListAndJoin.cpp:
```diff
@@ -7,6 +7,7 @@
 #include <iostream>
 #include <memory>
 #include <unordered_set>
+#include <cassert>
```
Please sort headers, there is a script for this.

tarun018 replied:
Updated.

tfar commented on Swiften/MIX/MIXRegistry.h:
```diff
@@ -8,6 +8,7 @@
 
 #include <memory>
 #include <unordered_set>
+#include <string>
```
Please sort headers.

tarun018 replied:
Updated.

tfar commented on Swiften/MIX/MIXRegistry.cpp:
```diff
@@ -0,0 +1,102 @@
+namespace Swift {
+
+MIXRegistry::MIXRegistry(const JID& ownJID, IQRouter* iqRouter, XMPPRoster* xmppRoster) : ownJID_(ownJID), iqRouter_(iqRouter), xmppRoster_(xmppRoster) {
+    scopedConnection_ = xmppRoster_->onInitialRosterPopulated.connect(boost::bind(&MIXRegistry::syncRegistryWithRoster, this));
```
`scopedConnection_` doesn't sensibly describe the variable. That it is a scoped connection is clear from the type. Maybe name it `initialRosterPopulationConnection_` or something like that

tarun018 replied:
Updated.

tfar commented on Swiften/MIX/MIXRegistry.cpp:
```diff
@@ -0,0 +1,102 @@
+
+}
+
+void MIXRegistry::syncRegistryWithRoster() {
```
This can be a C++11 lambda instead in the ctor, as the method name doesn't match project convention anyway (`handle*`).

tarun018 replied:
Added a lambda function.

tfar commented on Swiften/MIX/MIXRegistry.cpp:
```diff
@@ -0,0 +1,102 @@
+    SWIFT_LOG(debug) << "Syncing with roster items. " << std::endl;
+    for (auto item : xmppRoster_->getItems()) {
+        if (item.isMIXChannel()) {
+            SWIFT_LOG(debug) << item.getJID() << std::endl;
```
Pretty detailed logging. Probably not needed.

tarun018 replied:
Updated.

tfar commented on Swiften/MIX/MIXRegistry.cpp:
```diff
@@ -0,0 +1,102 @@
+            SWIFT_LOG(debug) << item.getJID() << std::endl;
+            auto i = entries_.find(item.getJID());
+            if (i == entries_.end()) {
+                SWIFT_LOG(debug) << "Adding " << item.getJID() << std::endl;
```
Pretty detailed logging. Probably not needed.

tarun018 replied:
Updated.

tfar commented on Swiften/MIX/MIXRegistry.h:
```diff
@@ -0,0 +1,83 @@
+            void handleJoinResponse(MIXJoin::ref, ErrorPayload::ref);
+            void handleLeaveResponse(MIXLeave::ref, ErrorPayload::ref);
+
+        public:
```
Please move those in front of the the private methods above.

tarun018 replied:
Updated.

tfar commented:
@Kev, both @intosi  and I have reviewed this PR. If you can give it a review too I can go on and submit it to master.
